mlir/lib/PauliFrame/Transforms/CliffordTToPauliFramePatterns.cpp

```
@@ -0,0 +1,502 @@
+
+    FlushOp flushOp = rewriter.create<FlushOp>(loc, rewriter.getI1Type(), rewriter.getI1Type(),
+                                               outQubitType, inQubits[0]);
+
+    auto pauliZOutQubit = insertPauliOpsAfterFlush(rewriter, loc, flushOp);
+
+    op->setOperands(pauliZOutQubit);
+
+    return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Clifford+T to Pauli Frame Patterns
+//===----------------------------------------------------------------------===//
+
+/**
+ * @brief Rewrite pattern for Clifford+T ops -> PauliFrame
+ */
+struct CliffordTToPauliFramePattern : public OpRewritePattern<CustomOp> {
+    using OpRewritePattern::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(CustomOp op, PatternRewriter &rewriter) const override
+    {
+        auto op_enum = hashGate(op);
+        switch (op_enum) {
+        case GateEnum::I:
+            return convertPauliGate(op, rewriter, false, false);
+        case GateEnum::X:
+            return convertPauliGate(op, rewriter, true, false);
+        case GateEnum::Y:
+            return convertPauliGate(op, rewriter, true, true);
+        case GateEnum::Z:
+            return convertPauliGate(op, rewriter, false, true);
+        case GateEnum::H:
+            return convertCliffordGate(op, rewriter, CliffordGate::Hadamard);
+        case GateEnum::S:
+            return convertCliffordGate(op, rewriter, CliffordGate::S);
+        case GateEnum::CNOT:
+            return convertCliffordGate(op, rewriter, CliffordGate::CNOT);
+        case GateEnum::T:
+            return convertNonCliffordGate(op, rewriter);
+        case GateEnum::Unknown: {
+            op->emitError() << "Unsupported gate: '" << op.getGateName()
+                            << "'. Only Clifford+T gates are supported for Pauli frame conversion: "
+                            << "I, X, Y, Z, H, S, S†, T, T†, and CNOT";
+            return failure();
```

I'm not sure if that's your intention, but this will not stop the compilation process. The error will be printed, and a match failure returned (which skips the pattern), but otherwise everything will go on as normal.

---

Good catch, thanks. Instead of returning `failure()` here I've let it fall back to `llvm_unreachable` to abort: b94fd23c8

There's probably a better way to do this, e.g. if it comes across an unsupported gate then in principle it could flush the Pauli record for the qubits acted upon by that gate, keep the unsupported gate in the IR and carry on. For the sake of the prototype, where we can assume the initial program representation is in Clifford+T, I think it's fastest, easiest and safest to just abort.

---

Yes I think in general it's best to design passes such that they don't error if they can't do their thing, unless it's absolutely required. Regarding the failure mechanism, unfortunately we are quite constrained (probably intentionally, but I'm not sure of the reasons), within MLIR rewrite patterns:
- for one, LLVM disables exceptions across the board, so we cannot raise them where we wish
- then we have asserts, but those could be disabled via build flags, and already have a different purpose anyway
- MLIR does provide a `signalPassFailure` function, which aborts the current pipeline, and in our setup will abort the compiler, but it can only be used from within a pass, not a pattern

I'm not sure we've ever figured out the best (or intended) way to handle this 😅